skimage/filters/_gabor.py

```diff
@@ -72,35 +254,63 @@ def gabor_kernel(frequency, theta=0, bandwidth=1, sigma_x=None, sigma_y=None,
     >>> io.imshow(gk.real)  # doctest: +SKIP
     >>> io.show()           # doctest: +SKIP
     """
-    if sigma_x is None:
-        sigma_x = _sigma_prefactor(bandwidth) / frequency
-    if sigma_y is None:
-        sigma_y = _sigma_prefactor(bandwidth) / frequency
+    # handle deprecation
+    message = ('Using deprecated, 2D-only interface to gabor_kernel. '
+               'This interface will be removed in scikit-image 0.16. Use '
+               'gabor_kernel(frequency, sigma=(sigma_x, sigma_y)).')
+
+    if sigma_y is not None:
+        warn(message)
+        if 'sigma_x' in kwargs:
+            sigma = (sigma_y, kwargs['sigma_x'])
+        else:
+            sigma = (sigma_y, sigma)
+
+    # handle translation
+    if not isinstance(theta, coll.Iterable):
+        theta = (theta,) * (ndim - 1)
 
-    x0 = np.ceil(max(np.abs(n_stds * sigma_x * np.cos(theta)),
-                     np.abs(n_stds * sigma_y * np.sin(theta)), 1))
-    y0 = np.ceil(max(np.abs(n_stds * sigma_y * np.cos(theta)),
-                     np.abs(n_stds * sigma_x * np.sin(theta)), 1))
-    y, x = np.mgrid[-y0:y0 + 1, -x0:x0 + 1]
+    if not isinstance(sigma, coll.Iterable):
+        sigma = np.array([sigma] * ndim)
+    else:
+        sigma = np.append(sigma, [None] * (ndim - len(sigma)))
+    default_sigma = _sigma_prefactor(bandwidth) / frequency
+    sigma[(sigma == None).nonzero()] = default_sigma  # noqa
+    sigma = sigma.astype(None)
 
-    rotx = x * np.cos(theta) + y * np.sin(theta)
-    roty = -x * np.sin(theta) + y * np.cos(theta)
+    coords = _decompose_quasipolar_coords(1, theta)
+    base_axis = np.zeros(ndim)
+    base_axis[leading_axis] = 1
+    rot = _rotation(base_axis, coords)
 
-    g = np.zeros(y.shape, dtype=np.complex)
-    g[:] = np.exp(-0.5 * (rotx ** 2 / sigma_x ** 2 + roty ** 2 / sigma_y ** 2))
-    g /= 2 * np.pi * sigma_x * sigma_y
-    g *= np.exp(1j * (2 * np.pi * frequency * rotx + offset))
+    # calculate & rotate kernel size
+    spatial_dims = np.ceil(np.max(np.abs(n_stds * sigma * rot), axis=-1))
+    spatial_dims[spatial_dims < 1] = 1
+
+    # create mesh grid
+    m = np.mgrid.__getitem__([slice(-c, c + 1) for c in spatial_dims])
+
+    rotm = np.matmul(m.T, rot).T
+
+    gauss = _gaussian(rotm, center=0, sigma=sigma, ndim=ndim)
+
+    compm = np.matmul(m.T, frequency * coords).T
+
+    # complex harmonic function
+    harmonic = np.exp(1j * (2 * np.pi * compm.sum() + offset))
+
+    g = np.zeros(m[0].shape, dtype=np.complex)
+    g[:] = gauss * harmonic
 
     return g
 
 
-def gabor(image, frequency, theta=0, bandwidth=1, sigma_x=None,
-          sigma_y=None, n_stds=3, offset=0, mode='reflect', cval=0):
+def gabor(image, frequency=None, theta=0, bandwidth=1, sigma=None, sigma_y=None,
+          n_stds=3, offset=0, mode='reflect', cval=0, kernel=None, **kwargs):
```

Comment:
You need to add `kernel` to the kwargs list in the docstring. Explain there that kernels can be computed with the `gabor_kernel` function, and that the operation will be significantly faster if you precompute the kernel when applying the same filter to many images.

Reply:
(btw: how much faster? Do you have some "typical" numbers?)

Reply:
@jni I don't have number for right now, it's mostly on the theoretical side. The `gabor_kernel` function has a complexity of O(N<sup>5.6356</sup>), meaning in higher dimensional cases, the generation of the kernel could be very costly.